grpc.go

```diff
@@ -225,30 +225,47 @@ func (a *agentGRPC) onlineCPUMem(req *pb.OnlineCPUMemRequest) error {
 	// all containers an update each cpuset cgroup. This is not required in docker
 	// containers since they don't hot add/remove CPUs.
 	for _, c := range a.sandbox.containers {
-		agentLog.WithField("container", c.container.ID()).Debug("updating cpuset cgroup")
-		contConfig := c.container.Config()
-		cgroupPath := contConfig.Cgroups.Path
-
-		// In order to avoid issues updating the container cpuset cgroup, its cpuset cgroup *parents*
-		// MUST BE updated, otherwise we'll get next errors:
-		// - write /sys/fs/cgroup/cpuset/XXXXX/cpuset.cpus: permission denied
-		// - write /sys/fs/cgroup/cpuset/XXXXX/cpuset.cpus: device or resource busy
-		// NOTE: updating container cpuset cgroup *parents* won't affect container cpuset cgroup, for example if container cpuset cgroup has "0"
-		// and its cpuset cgroup *parents* have "0-5", the container will be able to use only the CPU 0.
-		if contConfig.Cgroups.Resources.CpusetCpus != "" {
-			agentLog.WithField("cpuset", contConfig.Cgroups.Resources.CpusetCpus).Debug("updating container cpuset cgroup parents")
-			// remove container cgroup directory
-			cgroupPath = filepath.Dir(cgroupPath)
-		}
-
-		if err := updateContainerCpuset(cgroupPath, connectedCpus, cookies); err != nil {
+		err := updateContainerCpuset(c, connectedCpus, cookies)
+		if err != nil {
 			return handleError(req.Wait, err)
 		}
 	}
 
 	return nil
 }
 
+func updateContainerCpuset(c *container, connectedCpus string, cookies cookie) error {
+	if c == nil {
```

Comment:
I guess the race condition is because of sanbox contains nil containers, but now the question is why sandbox contains nil containers? :thinking:

Reply:
@devimc the line that fails is https://github.com/kata-containers/agent/blob/master/grpc.go#L228
So my conclusionis that c.container.ID() called when is nil, looking at the code when `a.sandbox.containers` gets a new element, this is not fully created that is the reason I added https://github.com/kata-containers/agent/pull/412/commits/da9e9f6cb61b7d208540f9ea65ee747548bb2e7f this help to not add the contaiener to the list until has all the values we want like c.container field.

Reply:
@jcvenegas ohh great!, then these conditions are no more needed? Could you confirm?

Reply:
Ditto @devimc. We'd better guarantee the container list doesn't have nil pointers instead of trying to catch it everywhere.